### ISSUE TITLE
Color edit marked as changed without actual changes

### ISSUE THREAD

**Issue by GamingMinds-DanielC** [ocornut/imgui#6722]
**Version/Branch of Dear ImGui:**

Version: 1.89.8 WIP (18974)
Branch: docking

**Back-end/Renderer/Compiler/OS**

Back-ends: Win32 + DX11
Compiler: VS 2019 / 2022
Operating System: Win 10

**My Issue/Question:**

I used the color widget as a reference for a custom widget with mode switching and noticed an unwanted behavior that I then tried with the original color widget. Whenever the display mode is changed or the value is copied into the clipboard, the widget is marked as "deactivated after edit".

**Standalone, minimal, complete and verifiable example:**

```c++
		static float color[ 4 ] = { 1.0f, 1.0f, 1.0f, 1.0f };
		static int   counter    = 0;

		ImGui::Text(
		    "Right click on the widget to open the color edit options,\n"
		    "then change the display mode or copy the value into the\n"
		    "clipboard. Notice how the counter is increased." );

		ImGui::ColorEdit4( "Color", color );
		if ( ImGui::IsItemDeactivatedAfterEdit() )
			counter++;

		ImGui::Text( "Deactivated after Edit: %i", counter );
```

**Source of the problem and possible solution:**

The popup is first displayed and later opened with constant label "context" within the group that defines the widget, so any action within the popup affects the flags of the group.

In my custom widget, I solved this by displaying (if open) the popup before starting the group. To not have ID conflicts, I surrounded id with `PushID("context")` and `PopID()` and gave the popup itself the label of the widget as ID. I deferred the opening of the popup from within the group to after it, surrounding it with the same "context" ID again. This solved the issue for my custom widget, a change of display mode no longer marks the widget as having been edited. Maybe this fix is feasible for the color widget as well.

**Comment by GamingMinds-DanielC** [comment]
Found a rather elegant solution, pull request incoming...

**Comment by ocornut** [comment]
Hello,

The checkbox/selectables inside those popups are calling `MarkItemEdited()` indeed.

It seems like you are introducing a very large feature to fix this specific case.
As we know `ColorEditOptionsPopup()` doesn't perform any edit, here's my suggestion:

[imgui-57ef206-WIP LockMarkEdited (#6722).patch](https://github.com/ocornut/imgui/files/12443439/imgui-57ef206-WIP.LockMarkEdited.6722.patch)

This would also conveniently allow removing the small amount of code using `ImGuiInputTextFlags_NoMarkEdited`.

What do you think?

**Comment by ocornut** [comment]
In principles I can see that #6731 provides a wider feature, but I cannot see a real use for it presently, other than solving this very specific problem you are having which can be solved with those ~6 lines of code changes, so unless you see an issue this situation I'd be always tempted to pick the simplest one.

**Comment by GamingMinds-DanielC** [comment]
Your version sure is a lot simpler in terms of code changes. It leaves the activation, deactivation and hovering reports in, but the main issue (deactivated after edit) should be solved*. Well, after adding two more lines in `ColorPickerOptionsPopup` (that one has the same problem as `ColorEditOptionsPopup`) it will be. That is already enough for my purposes (not filling up the undo stack with non-edits).

\* Not tested yet since I'm not at my office PC at weekends, just judging by looking at the patch.

**Comment by GamingMinds-DanielC** [comment]
I was just about to report back that I tested the patch and it works as expected. I did apply the locking to the picker popup as well, exactly as you did in the actual commit just now.

> As a tangential comment, if you use a `InputInt()` to type one value without pressing Enter, then edit it back to initial value and deactivate you'd face the same issue, and I believe an Undo system should be able to "discard" the non-edit based on actual data.

Usually yes, but in some cases having an undo for explicitly edited values even if they didn't change can be expected and can make some things easier (f.e. with automation that makes a "change", exports some data and undoes the change). Just like pushing a style on a stack, using it and popping it again is easier than checking if the style actually changes and making the push and pop conditional.
